src/main/java/hudson/remoting/FastPipedInputStream.java

```diff
@@ -53,11 +54,53 @@ public class FastPipedInputStream extends InputStream {
 
     private final Throwable allocatedAt = new Throwable();
 
+    private static final int DEFAULT_BUFFER_SIZE = 0x10000;
+    private static final Cleaner CLEANER = Cleaner.create();
+
+    /**
+     * Cleanup action for resource management that avoids reference to the outer
+     * class instance.
+     */
+    private static final class CleanAction implements Runnable {
+        private final byte[] buffer;
+        private final WeakReference<FastPipedInputStream> streamRef;
+
+        CleanAction(FastPipedInputStream stream) {
+            this.buffer = stream.buffer;
+            this.streamRef = new WeakReference<>(stream);
+        }
+
+        @Override
+        public void run() {
+            FastPipedInputStream stream = streamRef.get();
+            if (stream == null) {
+                return;
+            }
+
+            synchronized (buffer) {
+                if (stream.source != null && stream.closed == null) {
+                    stream.closed = new ClosedBy(null);
+                    // Release any pending writers.
+                    buffer.notifyAll();
```

Comment:
Seems like you could just call `stream.close()` here rather than duplicating code?

I think you should not be holding even the `WeakReference<FastPipedInputStream>` here, meaning that all you actually needed to do in the cleaner was `buffer.notifyAll`; the `stream.closed` value presumably does not matter if the `stream` is gone.